### PR TITLE
User documentation for replica scheduling preferences

### PR DESCRIPTION
**Pull request by irfanurrehman**

### CODE REVIEW

File: docs/userguide.md [hunk 1]
```diff
@@ -387,3 +387,164 @@ employed by deployment:
...
+      weight: 1
+    B:
+      weight: 2
+```
+A gets 3 and B gets 6 replicas in the proportion of 1:2. C does not get 
+any replica as missing weight preference is considered as weight=0.
+
+#### Distribute replicas in weighted proportions, also enforcing replica limits per cluster
+
+```bash
```
Comment (gyliu513) [inline docs/userguide.md]: Would you please show more detail for how does the `minReplicas` will impact the placement policy? I created a policy as this:

```
apiVersion: scheduling.federation.k8s.io/v1alpha1
kind: ReplicaSchedulingPreference
metadata:
  name: test-deployment
  namespace: test-namespace
spec:
  targetKind: FederatedDeployment
  totalReplicas: 10
  rebalance: true
  clusters:
   cluster1:
   - minReplicas: 2
     maxReplicas: 6
     weight: 2
   cluster2:
   - minReplicas: 4
     maxReplicas: 8
     weight: 3
```

And the result is cluster1 has 3 and cluster2 has 7, and my expectation is cluster1 should have 4 while cluster2 should have 6.

File: docs/userguide.md [hunk 2]
```diff
@@ -387,3 +387,164 @@ employed by deployment:
 ```bash
 NAMESPACED=y FEDERATION_NAMESPACE=<namespace> ./scripts/delete-federation.sh
 ```
+
+## Higher order behaviour
```
Comment (gyliu513) [inline docs/userguide.md]: How about `Advanced Features`

File: docs/userguide.md (hunk 1 above)
Comment (quinton-hoole-2) [reply docs/userguide.md]: Yes, that looks like a bug.  Your expectation is the same as mine.  Please file an issue.

Comment (gyliu513) [reply docs/userguide.md]: Thanks @quinton-hoole-2 , opened an issue at https://github.com/kubernetes-sigs/federation-v2/issues/319 to trace.

File: docs/userguide.md [hunk 3]
```diff
@@ -387,3 +387,164 @@ employed by deployment:
...
+
+The architecture of federation v2 API allows higher level APIs to be 
+constructed using the mechanics provided by the core API types (`template`, 
+`placement` and `override`) and associated controllers for a given resource.
+Further sections describe few of higher level APIs implemented as part of 
+Federation V2.
+
+###  ReplicaSchedulingPreference
+
+ReplicaSchedulingPreference provides an automated mechanism of distributing
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply docs/userguide.md]: Updated!

File: docs/userguide.md [hunk 4]
```diff
@@ -387,3 +387,164 @@ employed by deployment:
...
+These also include semantics to allow redistribution of replicas dynamically 
+in case some replica pods remain unscheduled in some clusters, for example 
+due to insufficient resources in that cluster. 
+
+RSP is used in place of ReplicaSchedulingPreference for brevity in text further on.
+
+The RSP controller works in a sync loop observing the RSP resource and the 
+matching `namespace/name` pair `FederatedDeployment` or `FederatedReplicaset` 
+resource. 
+If it finds that both RSP and its target template resource, the type of which
```
[the remark this answers is not in the document]
Comment (irfanurrehman) [reply docs/userguide.md]: This is written more as a paragraph then an itemised algorithm, segregating only these 2 ifs is not logical keeping rest of the paragraph same. Skipping this. Let me know if you have concerns.

File: docs/userguide.md (hunk 1 above)
Comment (irfanurrehman) [reply docs/userguide.md]: I think you have already raised it as an issue, lets track this as a bug there.

File: docs/userguide.md (hunk 2 above)
Comment (irfanurrehman) [reply docs/userguide.md]: The current title seems apt to me, given that this section intends to provide details about the higher level API and associated controllers.

**Comment by irfanurrehman** [comment]
@gyliu513, have updated, please take a look.
@pmorie you can approve post @gyliu513's lgtm.

**Pull request merged**